### PR TITLE
Stop computing kername hashes lazily.

### PR DESCRIPTION
**Pull request by ppedrot**

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

File: kernel/names.ml [hunk 1]
```diff
@@ -368,14 +368,19 @@ module KerName = struct
...
   }
 
   type kernel_name = t
 
   let make modpath knlabel =
-    { modpath; knlabel; refhash = -1; }
+    let open Hashset.Combine in
+    let refhash = combine (ModPath.hash modpath) (Label.hash knlabel) in
+    (* Ensure positivity on all platforms. *)
+    let refhash = refhash land 0x3FFFFFFF in
```
Comment (SkySkimmer) [inline kernel/names.ml:381]: We don't need to ensure positivity if we're not using `-1` as a sentinel do we?

Comment (ppedrot) [reply kernel/names.ml:381]: I think the comment is misleading, it's there not so much for positivity but for portability between platforms. At some point we tried to remove the sources of divergence for vo files between 32 and 64-bit architectures, but AFAIK we've given up on that.

**Comment by ppedrot** [comment]
This doesn't seem bad. From the CPU cycles there is a small something going on with unimath, hott and category-theory but it's not reflected at the macroscopic level. Probably something related to universes judging from the kind of development affected. I'll have a look at the worse unimath slowdowns but honestly this could be merged as-is already.

**Comment by ppedrot** [comment]
Actually, the problem is quite trivial. The slowdowns are due to the call to `RedFlags.red_projection` which repeatedly rebuilds a constant out of a projection when performing conversion. I guess we could just store the constant directly in the projection to prevent this. Does it look OK to you @SkySkimmer ?

**Comment by SkySkimmer** [comment]
Sounds ok

**Comment by ppedrot** [comment]
En passant, I find extremely suspect that we rely so heavily on the compat layer at the kernel level. We should have two different entries in the transparent state, one for the constant and one for the primitive projection, and keep them synchronized (for now) in the upper layers.

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
This looks OK now.

**Comment by ppedrot** [comment]
@SkySkimmer this should be ready. In the last iteration I've also removed the positivity-enforcing code, it's useless now.

**Review by SkySkimmer — APPROVED** [review]

**Comment by ppedrot** [comment]
What's with the test-suite failure? Was positivity enforcement actually needed or did we depend on the specific hash order for printing?

**Comment by SkySkimmer** [comment]
Search uses hashset fold so its printing order does depend on hashes

**Comment by ppedrot** [comment]
Meh. Should I fix the test-suite or keep the old positivization hack?

**Comment by SkySkimmer** [comment]
Fix the tests (and probably dpdgraph will need a fix too) IMO

**Comment by ppedrot** [comment]
Finally I reinstated the truncation, I find this to be better for backwards compat.

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**